nornir/plugins/inventory/netbox.py

```diff
@@ -143,10 +147,66 @@ def __init__(
             "NB_TOKEN", "0123456789abcdef0123456789abcdef01234567"
         )
 
+        # Load in user private key from environment var or file load
+        nb_private_key = nb_private_key or os.environ.get("NB_PRIVATE_KEY", None)
+
+        if os.environ.get('NB_PRIVATE_KEY') is not None:
+            nb_private_key = os.environ.get('NB_PRIVATE_KEY')
+        elif os.environ.get('NB_PRIVATE_KEY') is None and nb_private_key:
+            with open(nb_private_key, 'r') as f:
+                nb_private_key = f.read()
+
         session = requests.Session()
         session.headers.update({"Authorization": f"Token {nb_token}"})
         session.verify = ssl_verify
 
+        # Fetch all secrets from Netbox
+        # Since the api uses pagination we have to fetch until no next is provided
+        if nb_private_key:
+            # Get session key first in order to get decrypted secrets
+
+            url = f"{nb_url}/api/secrets/get-session-key/"
+            nb_session_key: str = None
+
+            headers = {
+                'Connection': 'keep-alive',
+                'Accept': 'application/json, text/javascript, */*; q=0.01',
+                'DNT': '1',
+                'X-Requested-With': 'XMLHttpRequest',
+                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
+                'Accept-Encoding': 'gzip, deflate',
+                'Accept-Language': 'en-US,en;q=0.9',
+            }
+
+            payload = {
+                "private_key": nb_private_key
+            }
+            r = session.post(url, headers=headers, data=payload)
+
+            if not r.status_code == 200:
+                raise ValueError(f"Failed to get session key from Netbox instance {nb_url}")
+
+            # Add session key to cookie jar for session
+            session_key_cookie = r.json()
+            session.cookies.set(name="session_key", value=session_key_cookie.get("session_key"))
+
+
+            # Get all secrets
+
+            url = f"{nb_url}/api/secrets/secrets/?limit=0"
+            nb_secrets: List[Dict[str, Any]] = []
+
+            while url:
```

Comment:
We should probably put this logic (url paging) in it's own function, since we are repeating the logic 2 times (getting devices & secrets)

Reply:
I put it in a class method. It is working.